Firestore/core/src/index/index_byte_encoder.h

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FIRESTORE_CORE_SRC_INDEX_INDEX_BYTE_ENCODER_H_
+#define FIRESTORE_CORE_SRC_INDEX_INDEX_BYTE_ENCODER_H_
+
+#include <memory>
+#include <string>
+
+#include "Firestore/core/src/model/field_index.h"
+#include "Firestore/core/src/nanopb/byte_string.h"
+#include "Firestore/core/src/nanopb/nanopb_util.h"
+#include "Firestore/core/src/util/ordered_code.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/string_view.h"
+
+namespace firebase {
+namespace firestore {
+namespace index {
+
+/** An index value encoder. */
+class DirectionalIndexByteEncoder {
+ public:
+  virtual void WriteBytes(pb_bytes_array_t* val) = 0;
+
+  virtual void WriteString(absl::string_view val) = 0;
+
+  virtual void WriteLong(int64_t val) = 0;
+
+  virtual void WriteDouble(double val) = 0;
+
+  virtual void WriteInfinity() = 0;
+
+  virtual ~DirectionalIndexByteEncoder() = default;
```

Comment:
Should this be at the top? My gut tells me it should be constructor/desctructor followed by custom methods.